File: MasterNode/src/main/java/com/yahoo/sdvornik/server/BroadcastServer.java
package com.yahoo.sdvornik.server;

import com.yahoo.sdvornik.sharable.Constants;
import com.yahoo.sdvornik.sharable.BroadcastMessage;
import com.yahoo.sdvornik.broadcaster.BroadcastMessageEncoder;
import com.yahoo.sdvornik.main.EntryPoint;
import io.netty.bootstrap.Bootstrap;
import io.netty.channel.*;
import io.netty.channel.socket.nio.NioDatagramChannel;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.TimeUnit;

public class BroadcastServer {

    private static final Logger log = LoggerFactory.getLogger(BroadcastServer.class.getName());

    private final EventLoopGroup udpEventLoopGroup;

    private BroadcastMessage broadcastMessage;

    public BroadcastServer(EventLoopGroup udpEventLoopGroup, BroadcastMessage broadcastMessage) {
        this.udpEventLoopGroup = udpEventLoopGroup;
        this.broadcastMessage = broadcastMessage;
    }

    public void init() {
        final Bootstrap udpBootstrap = new Bootstrap();

        udpBootstrap.group(udpEventLoopGroup)
                .channel(NioDatagramChannel.class)
                .option(ChannelOption.SO_BROADCAST, true)
                .handler(new BroadcastMessageEncoder());


        ChannelFuture udpFuture = udpBootstrap.bind(0);
        udpFuture.addListener(new ChannelFutureListener() {
            @Override
            public void operationComplete(ChannelFuture future) throws Exception {
                if(!future.isSuccess()) {
                    log.error("Can't run Broadcast server", future.cause());
                    EntryPoint.stop();
                    return;
                }
                log.info("Successfully init Broadcast server");
                final Channel udpChannel = future.channel();
                udpEventLoopGroup.scheduleAtFixedRate(
                        new Runnable() {
                            @Override
                            public void run() {
                                udpChannel.writeAndFlush(broadcastMessage);
                                //log.info("Successfully broadcast message");
                            }
                        },
                        0,
                        Constants.BROADCAST_INTERVAL_IN_MS,
                        TimeUnit.MILLISECONDS
                );
            }
        });
    }
}

File: master/src/main/kotlin/com/yahoo/sdvornik/server/BroadcastServer.kt
package com.yahoo.sdvornik.server

import com.yahoo.sdvornik.Constants
import com.yahoo.sdvornik.main.MasterEntryPoint
import com.yahoo.sdvornik.message.BroadcastMessage
import com.yahoo.sdvornik.message.codec.BroadcastMsgCodec
import com.yahoo.sdvornik.utils.Utils
import io.netty.bootstrap.Bootstrap
import io.netty.channel.ChannelFuture
import io.netty.channel.ChannelFutureListener
import io.netty.channel.ChannelOption
import io.netty.channel.EventLoopGroup
import io.netty.channel.socket.nio.NioDatagramChannel
import org.slf4j.LoggerFactory
import java.util.concurrent.TimeUnit

/**
 * @author Serg Dvornik <sdvornik@yahoo.com>
 */
/**
 * UDP broadcasting server
 */
class BroadcastServer
/**
 * Ctor.
 */
(private val udpEventLoopGroup: EventLoopGroup) {


    fun init() {
        val udpBootstrap = Bootstrap()

        udpBootstrap.group(udpEventLoopGroup)
                .channel(NioDatagramChannel::class.java)
                .option(ChannelOption.SO_BROADCAST, true)
                .handler(BroadcastMsgCodec())


        val udpFuture = udpBootstrap.bind(0)
        udpFuture.addListener(ChannelFutureListener { future ->
            if (!future.isSuccess) {
                log.error("Can't run Broadcast server", future.cause())
                MasterEntryPoint.INSTANCE.stop()
                return@ChannelFutureListener
            }
            log.info("Successfully init Broadcast server")
            val udpChannel = future.channel()
            udpEventLoopGroup.scheduleAtFixedRate(
                    object : Runnable {
                        internal val broadcastMessage = BroadcastMessage(Utils.getLocalHostAddress(), Constants.PORT)
                        override fun run() {
                            udpChannel.writeAndFlush(broadcastMessage)
                        }
                    },
                    0,
                    Constants.BROADCAST_INTERVAL_IN_MS.toLong(),
                    TimeUnit.MILLISECONDS
            )
        })
    }

    companion object {

        private val log = LoggerFactory.getLogger(BroadcastServer::class.java)
    }
}